### DISCUSSION TITLE
Prevent Next.js build process from converting RGBA to HSLA

### DISCUSSION THREAD

**Discussion by NabeelAhmed1721** [vercel/next.js#41288] [Help]
I have a component where I use a CSS Module stylesheet. Every style is working fine except for the background color, where when I build my project, Next.js converts my

`background-color: rgba(17,17,17,0.5)` to `background-color: hsla(0,0%,7%,.5)`

I wouldn't have cared if it did convert to HSLA, but what happens is the HSLA value is not equal to `rgba(17,17,17,0.5)`. It is actually `rgba(18,18,18,0.5)`, which creates a weird color shift on the webpage.

> note: this only occurs when I build. It doesn't convert the color format on a development server

I don't know why it is doing this. I have tried editing the `next.config.js` file but to no luck.

Thanks.

**Comment by icyJoseph** [comment]
I don't have a solution yet, but the `postcss` loader is running a `cssnano` optimisation called `colormin`, which I also see misbehaving. I've tried with a `postcss` config file, but no luck, I guess there's a hole WebPack Plugin that optimizes CSS, which we need to override in `next.config.js`.

**Reply by icyJoseph** [reply]
@NabeelAhmed1721 I've dived through the code and I am fairly sure, unless I am missing some magic config, that it is not possible to turn this off, at the moment anyway, because of this line https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/plugins/css-minimizer-plugin.ts#L46:

```js
return postcss([cssnanoSimple({}, postcss)])
```

Where the special plugin `cssnano-simple` is initiated without options. If I pass `{ colormin: {exclude: true}}` then I finally manage to turn it off!

## Buying into a potential problem

There is one way out. You could on your `next.config.js`'s webpack key, remove the minimizer that's causing this issue. Unfortunately, it is not easy to tell which one it is. In the version I am using, it is the one at index 1.

```js
 config.optimization.minimizer.length = 1;
```

The above would do the trick, but it'd disable all of the CSS minification. 

Alternatively you could define your own minifier, by extending the `CssMinimizerPlugin` from Next.js:

```js
const cssnanoSimple = require("next/dist/compiled/cssnano-simple");
const postcss = require("postcss");
const {
  CssMinimizerPlugin,
} = require("next/dist/build/webpack/plugins/css-minimizer-plugin");

const { sources } = require("next/dist/compiled/webpack/webpack");

class CustomCssMinizer extends CssMinimizerPlugin {
  optimizeAsset(file, asset) {
    const postcssOptions = {
      ...this.options.postcssOptions,
      to: file,
      from: file,
    };

    let input;
    if (postcssOptions.map && asset.sourceAndMap) {
      const { source, map } = asset.sourceAndMap();
      input = source;
      postcssOptions.map.prev = map ? map : false;
    } else {
      input = asset.source();
    }

    return postcss([cssnanoSimple({ colormin: { exclude: false } }, postcss)])
      .process(input, postcssOptions)
      .then((res) => {
        if (res.map) {
          return new sources.SourceMapSource(res.css, file, res.map.toJSON());
        } else {
          return new sources.RawSource(res.css);
        }
      });
  }
}

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack(config, { nextRuntime }) {
    if (!nextRuntime) {
      const customCSSMinizer = (compiler) => {
        new CustomCssMinizer({
          postcssOptions: {
            map: {
              // `inline: false` generates the source map in a separate file.
              // Otherwise, the CSS file is needlessly large.
              inline: false,
              // `annotation: false` skips appending the `sourceMappingURL`
              // to the end of the CSS file. Webpack already handles this.
              annotation: false,
            },
          },
        }).apply(compiler);
      };

      config.optimization.minimizer[1] = customCSSMinizer;
    }
    return config;
  },
};

module.exports = nextConfig;
```

But now you are in charge of maintaining this, and figuring out where to put the minimizer:

```js
 config.optimization.minimizer[1] = customCSSMinizer;
```

This line becomes extremely sensitive! The challenge here is that we must remove the CSS minimizer and place our own. Not very nice I must say, and I don't see where there could be a configuration flag to change this. 

I am unfortunately short on time, and I haven't been able to test:

- https://github.com/vercel/next.js/blob/canary/packages/next/server/config-shared.ts#L571 (optimizeCSS flag)
- To see if the minizers can be identified via other means, such as casting to a string and then looking for the class name in the string
- Various cssnano.config file types and options, though I think because cssnano-simple is pre-compiled we can't really do anything about it. 
- PostCSS config file variations

Hope to hear from others! Hopefully we could put out an MR to allow us to pass config options to `cssnano-simple`!


> Note that `colormin: false` also does work in all cases listed above.

**Reply by NabeelAhmed1721** [reply]
Thanks for your help!

To add to your investigation, I changed my `next.config.js` to disable `optimizeCSS`, and I am still experiencing the same result.

Your solution to a custom `next.config.js` file works well, but, it is quite convoluted. Although not ideal, I think I'll just try using inline CSS for now as it seems to bypass the `colormin` optimization.

**Comment by Keekuun** [comment]
cssnano bug https://github.com/cssnano/cssnano/issues/1648
